test/sequence/token_refresh_sequence_test.rb

```diff
@@ -7,82 +7,162 @@
-                         'Authorization' =>
-                              "Basic #{Base64.strict_encode64(@instance.client_id +
-                                                                  ':' +
-                                                                  @instance.client_secret)}" },
-              body: { 'grant_type' => 'refresh_token',
-                      'refresh_token' => @instance.refresh_token })
-        .to_return(status: 200,
-                   body: @standalone_token_exchange.to_json,
-                   headers: { content_type: 'application/json; charset=UTF-8',
-                              cache_control: 'no-store',
-                              pragma: 'no-cache' })
-
-      # Responses must NOT contain client_id in the body or the client secret in any situation
-      stub_request(:post, @instance.oauth_token_endpoint)
-        .with(body: /client_id|client_secret/)
-        .to_return(status: 401)
+
+    headers = {
+      'Content-Type' => 'application/x-www-form-urlencoded'
+    }
+    body = {
+      'grant_type' => 'refresh_token',
+      'refresh_token' => @instance.refresh_token
+    }
+
+    body_with_scope = {
+      'grant_type' => 'refresh_token',
+      'refresh_token' => @instance.refresh_token,
+      'scope' => @instance.scopes
+    }
+
+    exchange_response = @standalone_token_exchange.dup
+
+    exchange_response['token_type'] = 'unknown' if failure_mode == :bad_token_type
+    exchange_response.delete('scope') if failure_mode == :no_scope
+    exchange_response.delete('access_token') if failure_mode == :no_access_token
+
+    if @instance.client_secret.present?
+      headers['Authorization'] = "Basic #{Base64.strict_encode64(@instance.client_id + ':' + @instance.client_secret)}"
     else
-      stub_request(:post, @instance.oauth_token_endpoint)
-        .with(headers: { 'Content-Type' => 'application/x-www-form-urlencoded' },
-              body: { 'client_id' => @instance.client_id,
-                      'grant_type' => 'refresh_token',
-                      'refresh_token' => @instance.refresh_token })
-        .to_return(status: 200,
-                   body: @standalone_token_exchange.to_json,
-                   headers: { content_type: 'application/json; charset=UTF-8',
-                              cache_control: 'no-store',
-                              pragma: 'no-cache' })
+      body['client_id'] = body_with_scope['client_id'] = @instance.client_id
     end
 
-    # To test rejection of invalid client_id
+    response_headers = { content_type: 'application/json; charset=UTF-8',
+                         cache_control: 'no-store',
+                         pragma: 'no-cache' }
+
+    response_headers.delete(:cache_control) if failure_mode == :cache_control_off
+    response_headers.delete(:pragma) if failure_mode == :pragma_off
+
+    exchange_response_json = exchange_response.to_json
+    exchange_response_json = '<bad>' if failure_mode == :bad_json_response
```

Comment:
I think it would be nice to group all of the `failure_mode` changes together.

Reply:
Consolidated all of these into one spot.